Automates AppStoreConnect submissions

ashfurrow opened the pull request:
This PR automates the work of submitting a build through AppStoreConnect. It includes #2747. It provides a new `make promote_beta_to_submission` command to promote the most recent beta to the App Store. It pushes to the `app_store_submission` branch, and CI looks for that branch to run `make promote_beta`, which works with Fastlane based on [these docs](https://github.com/fastlane/fastlane/pull/13882).

This PR is a WIP because it needs the following tasks completed:

- [ ] Need to make sure Fastlane sets the proper IDFA metadata.
- [x] Need to make sure `fastlane/metadata` directory is all up-to-date. (We have a bunch of stuff in there that's language-specific, but it's all English. I'm thinking we should remove it all, what do you think @alloy?) 
- [x] Need to use git tags to mark a build as submitted.
- [ ] Testing.

ArtsyOpenSource commented:
<!--
  0 Errors
  1 Warning: Needs testing on a Phone if ch...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Needs testing on a Phone if change is non-trivial</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_travis">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

ashfurrow commented:
Okay, this is ready to test. The easiest way to test is to try promoting a beta and see if it works, paying close attention to the IDFA setting used by `deliver`. The nice thing about this PR is that, if it doesn't work, we can still do it manually. 

I removed all the localized metadata because we don't have any translations, and I updated everything with new URLs and stuff. The release_notes.txt are unchanged and need modifying, I've noted that in the docs.

ashfurrow commented:
@orta going to assign this to you but if you'd prefer not to merge a deploy process-changing PR from a contributor who is about to go on vacation, I'd totally understand!

orta commented on fastlane/Fastfile:
```diff
@@ -119,3 +112,36 @@ lane :ship_beta do
+
+  puts "Tagging submission and pushing to GitHub."
+  tag_and_push(tag: "#{beta.train_version}-#{beta.build_version}-submission")
+  
```
this could probably just be the tag `2.5.1` for example, all betas are `2.5.1-0`, `2.5.1-1`

ashfurrow replied:
I wanted to keep this unique in the case that Apple rejects our beta and we need to resubmit with the same version (but different build numbers).

alloy commented:
> Need to make sure fastlane/metadata directory is all up-to-date. (We have a bunch of stuff in there that's language-specific, but it's all English. I'm thinking we should remove it all, what do you think @alloy?)

We used to have to submit multiple copies for various stores, but if that’s no longer the case then by all means 👍

ashfurrow commented:
@alloy yup, that's the case. 

During today's FrontEnd iOS Practice meeting, @orta and I discussed this and we're going to try this on our deploy next week. One TODO item was "test", and what better way to test than to actually try to deploy? We should pay attention to the IDFA setting that's set by Fastlane, but we can easily correct it if it's wrong (and I can follow up with a PR to Fastlane).

Okay, merging!
